### PR TITLE
#24155 Binary op make flag use_legacy optional except typecast and block format

### LINKED ISSUE #23310: As default, use binary_ng for width and block shard type when there is no broadcasting
Currently, if use_legacy is nullopt, binary_ng will delegate to legacy for width and block shard type when there is no broadcasting. Binary_ng support that so should not delegate.

### PR DESCRIPTION
**Pull request by dchenTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/24155)

### Problem description
Currently a lot of test cases require binary ops with flag use_legacy=False, because they only work with binary_ng. We need to add new heuristics as following to make sure without the flag, it still direct to binary_ng.

### What's changed
Add new heuristics to handle following cases. It also fixed https://github.com/tenstorrent/tt-metal/issues/23310 in the same time.

- Int type, int32, uint32, uint16, uint8
- Rank > 4
- Row_broadcast other than ADD, SUB, MUL
- Lhs_activations >1 or rhs_activiations is not empty (runtime error, sad path)
- Subtile broadcast type row_a_col_b, col_a_row_b

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/eltwise/test_binary_add_uint32.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ def test_binary_add_uint32_bcast(a_shape, b_shape, low_a, high_a, low_b, high_b,
         layout=ttnn.TILE_LAYOUT,
         memory_config=ttnn.DRAM_MEMORY_CONFIG,
     )
-    output_tensor = ttnn.add(input_tensor_a, input_tensor_b, use_legacy=False)
+    output_tensor = ttnn.add(input_tensor_a, input_tensor_b, use_legacy=None)
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/eltwise/test_binary_add_uint32.py:57]: If the defaults for use_legacy are already std::nullopt, is it possible to change all of these calls to not specify use_legacy?

E.g. for this line
```
output_tensor = ttnn.add(input_tensor_a, input_tensor_b)
```

Comment (dchenTT) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_add_uint32.py:57]: I can either do that in last commit, or do it in next PR (still work to do to totally make the flag optional). Having the flag make it easy to change and test between None and False.

Comment (bbradelTT) [reply tests/ttnn/unit_tests/operations/eltwise/test_binary_add_uint32.py:57]: Next PR is ok.

**Review by bbradelTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/test_binaryng_ND.py [hunk 2]
```diff
@@ -83,7 +83,7 @@ def test_ND_subtile_bcast(device, shapes, ttnn_fn):
         torch_input_tensor_b, layout=ttnn.TILE_LAYOUT, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG
     )
 
-    output_tensor = ttnn_fn(input_tensor_a, input_tensor_b, memory_config=ttnn.DRAM_MEMORY_CONFIG, use_legacy=False)
+    output_tensor = ttnn_fn(input_tensor_a, input_tensor_b, memory_config=ttnn.DRAM_MEMORY_CONFIG, use_legacy=None)
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/test_binaryng_ND.py:86]: Please consider renaming the test file, as this is no longer testing specifically binary_ng

File: tests/ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ def test_sub_fp32(device):
     z_torch = golden_fn(x_torch, y_torch)
     x_tt = ttnn.from_torch(x_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
     y_tt = ttnn.from_torch(y_torch, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
-    z_tt_sub = ttnn.sub(x_tt, y_tt, use_legacy=False)
+    z_tt_sub = ttnn.sub(x_tt, y_tt, use_legacy=None)
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py:18]: Same as above

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 4]
```diff
@@ -335,18 +401,20 @@ inline auto invoke_binary_ng(
...
-
-        if constexpr (requires { detail::preprocess_inputs(binary_op_type, lhs, rhs); }) {
-            auto [a, b] = detail::preprocess_inputs(binary_op_type, lhs, rhs);
-
-            return ttnn::prim::binary(
-                queue_id, a, b, binary_op_type, dtype, memory_config, output, activations, lhs_activation);
-        } else {
-            return ttnn::prim::binary(
-                queue_id, lhs, rhs, binary_op_type, dtype, memory_config, output, activations, lhs_activation);
+        if ((not detail::is_binary_ng_only(lhs, rhs, binary_op_type)) or *use_legacy) {
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: I'd prefer to avoid `if (...) { if (...) { ... } }` here, can you refactor this?

Note, the existing logic respects the value of `use_legacy` when it is set explicitly to `True` or `False` by the user, and does not use heuristics to determine which implementation to select. This is important for stability, regardless of whether the selected implementation is believed to support the configuration based on the heuristic. There are just too many combinations to test and we never want to lock a user out of a particular selection they request in case there's an edge-case we're not aware of.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 5]
```diff
@@ -514,7 +582,9 @@ Tensor RelationalBinary<binary_op_type>::invoke(
     const std::optional<bool>& use_legacy) {
     if (use_legacy ? *use_legacy
                    : binary::is_legacy_only(lhs, rhs, memory_config, output, lhs_activations, rhs_activations)) {
-        return detail::binary_impl(DefaultQueueId, binary_op_type, lhs, rhs, dtype, memory_config, output);
+        if ((not detail::is_binary_ng_only(lhs, rhs, binary_op_type)) or *use_legacy) {
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: Same as above.

Also, this logic probably needs to be propagated to the ops in `ttnn/cpp/ttnn/operations/eltwise/binary/device/binary_composite_op.cpp`, it looks like that was missed.

**Review by patrickroberts — CHANGES_REQUESTED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp (hunk 4 above)
Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: I added the binary_ng heuristics while respecting the original is_legacy_only(). Note there is "or *use_legacy" condition. So if use_legacy=True, it will skip binary_ng heuristics. I do realize I need to move it to the front of if condition.

Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: Refactored to following:
if (use_legacy ? *use_legacy
                   : binary::is_legacy_only(lhs, rhs, memory_config, output, lhs_activations, rhs_activations) and
                         (not detail::is_binary_ng_only(lhs, rhs, binary_op_type))) {

Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: Resolved in last commit.

File: tests/ttnn/unit_tests/operations/eltwise/test_binaryng_ND.py (hunk 2 above)
Comment (dchenTT) [reply tests/ttnn/unit_tests/operations/eltwise/test_binaryng_ND.py:86]: Future PR.

File: tests/ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py (hunk 3 above)
Comment (dchenTT) [reply tests/ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py:18]: Future PR.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp (hunk 5 above)
Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp]: Future PR. Put a note in https://github.com/tenstorrent/tt-metal/issues/24152

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 6]
```diff
@@ -282,6 +292,45 @@ inline auto any_uneven(const Tensor& a, const auto& b, const std::optional<Tenso
...
+        }
+        if (b.get_logical_shape()[-2] == 1 && a.get_logical_shape()[-2] > 1 && b.get_logical_shape()[-1] > 1 &&
+            a.get_logical_shape()[-1] == 1) {
+            return true;
+        }
+
+        if (any_row_broadcasted(a, b) and (is_block_format(a.get_dtype()) or is_block_format(b.get_dtype()))) {
+            // TODO
+            // return true;
+        }
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:329]: Would prefer to comment out the entire condition if the early return is to be excluded.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 7]
```diff
@@ -529,8 +579,11 @@ Tensor RelationalBinary<binary_op_type>::invoke(
     tt::stl::Span<const ttnn::operations::unary::UnaryWithParam> rhs_activations,
     const std::optional<bool>& use_legacy) {
     if (use_legacy ? *use_legacy
-                   : binary::is_legacy_only(lhs, rhs, memory_config, output, lhs_activations, rhs_activations)) {
-        return detail::binary_impl(DefaultQueueId, binary_op_type, lhs, rhs, dtype, memory_config, output);
+                   : binary::is_legacy_only(lhs, rhs, memory_config, output, lhs_activations, rhs_activations) and
+                         (not detail::is_binary_ng_only(lhs, rhs, binary_op_type))) {
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:583]: Thanks, I think this is a lot easier to understand the intent of than the previous iteration. 👍

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 8]
```diff
@@ -529,8 +579,11 @@ Tensor RelationalBinary<binary_op_type>::invoke(
...
+        {
+            return detail::binary_impl(DefaultQueueId, binary_op_type, lhs, rhs, dtype, memory_config, output);
+        }
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:586]: Remove extra braces?

File: tests/ttnn/unit_tests/gtests/tensor/test_tensor_nd_sharding.cpp [hunk 9]
```diff
@@ -223,9 +223,9 @@ TEST_P(NdShardingOpCompatTests, TestAdd) {
 
     auto sum_tensor = ttnn::add(tensor_a, tensor_b);
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/gtests/tensor/test_tensor_nd_sharding.cpp:224]: Point of clarification: which implementation is being selected here, binary_ng or legacy?

**Review by patrickroberts — APPROVED** [review]
Overall looks good, just one request for clarification, and some minor cleanup.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp (hunk 6 above)
Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:329]: yes, will clean up in future PR.

File: tests/ttnn/unit_tests/gtests/tensor/test_tensor_nd_sharding.cpp (hunk 9 above)
Comment (dchenTT) [reply tests/ttnn/unit_tests/gtests/tensor/test_tensor_nd_sharding.cpp:224]: Legacy, since it is uneven shard.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp (hunk 8 above)
Comment (dchenTT) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:586]: Oops. Will have plenty of chances to remove in later PRs
